src/QsCompiler/Transformations/ClassicallyControlledTransformation.cs

```diff
@@ -52,41 +53,83 @@ public static QsCompilation Apply(QsCompilation compilation)
+                    {
+                        return (true, literal, expr, scope, defaultScope);
+                    }
+                }
+            }
+
+            return (false, null, null, null, null);
+        }
+
         private ClassicallyControlledTransformation() { }
 
+        private class ClassicallyControlledSyntax : SyntaxTreeTransformation<ClassicallyControlledScope>
+        {
+            public ClassicallyControlledSyntax(ClassicallyControlledScope scope = null) : base(scope ?? new ClassicallyControlledScope()) { }
+
+            public override QsCallable onFunction(QsCallable c) => c; // Prevent anything in functions from being considered
+        }
+
         private class ClassicallyControlledScope : ScopeTransformation<NoExpressionTransformations>
         {
             public ClassicallyControlledScope(NoExpressionTransformations expr = null) : base(expr ?? new NoExpressionTransformations()) { }
 
-            private (bool, QsResult, TypedExpression, QsScope, QsScope) IsConditionedOnResultLiteralStatement(QsStatement statement)
-            {
-                if (statement.Statement is QsStatementKind.QsConditionalStatement cond)
-                {
-                    if (cond.Item.ConditionalBlocks.Length == 1 && (cond.Item.ConditionalBlocks[0].Item1.Expression is ExpressionKind.EQ expression))
-                    {
-                        var scope = cond.Item.ConditionalBlocks[0].Item2.Body;
-                        var defaultScope = cond.Item.Default.ValueOr(null)?.Body;
-
-                        if (expression.Item1.Expression is ExpressionKind.ResultLiteral exp1)
-                        {
-                            return (true, exp1.Item, expression.Item2, scope, defaultScope);
-                        }
-                        else if (expression.Item2.Expression is ExpressionKind.ResultLiteral exp2)
-                        {
-                            return (true, exp2.Item, expression.Item1, scope, defaultScope);
-                        }
-                    }
-                }
-
-                return (false, null, null, null, null);
-            }
-
+            private TypeArgsResolution GetCombinedType(TypeArgsResolution outer, TypeArgsResolution inner)
+            {
+                var result = new List<Tuple<QsQualifiedName, NonNullable<string>, ResolvedType>>();
+                var outerDict = outer.ToDictionary(x => (x.Item1, x.Item2), x => x.Item3);
+
+                return inner.Select(innerRes =>
+                {
+                    if (innerRes.Item3.Resolution is ResolvedTypeKind.TypeParameter typeParam &&
+                        outerDict.ContainsKey((typeParam.Item.Origin, typeParam.Item.TypeName)))
+                    {
+                        var outerRes = outerDict[(typeParam.Item.Origin, typeParam.Item.TypeName)];
+                        return Tuple.Create(innerRes.Item1, innerRes.Item2, outerRes);
+                    }
+                    else

```

Comment:
Isn't this missing to handle the case where the outer dictionary resolves type parameters that are not listed in the inner dictionary?